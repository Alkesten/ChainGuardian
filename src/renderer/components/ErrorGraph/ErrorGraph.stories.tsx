import * as React from 'react';
import { storiesOf } from '@storybook/react';
import { withKnobs, text, boolean, number, select } from '@storybook/addon-knobs';
import { ErrorGraph} from './ErrorGraph';

storiesOf('ErrorGraph', module).add('ErrorGraph', () => {
    
    return <ErrorGraph 
    getData={async ()=>{
        return(
            [2,5,9,12,1,5,6,8,10,10,11,5,7,8
            ]
        )
    }} />
}).addDecorator(withKnobs);
