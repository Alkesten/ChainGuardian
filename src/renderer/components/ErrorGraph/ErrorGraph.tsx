import * as React from "react";
import {useState, useEffect} from "react";
import { 
    BarChart, Bar, XAxis, Tooltip,
} from "recharts";

export interface IErrorGraphProps {
    getData: () => Promise<number[]>;
}

export const ErrorGraph: React.FunctionComponent<IErrorGraphProps> = (props: IErrorGraphProps) => {
    const [data, setData] = useState<Array<object>>([]);
    const [refreshIntervalId, setRefreshIntervalId] = useState<number>(0);
    const [lastRefreshTime,setLastRefreshTime] = useState<number>(0);

    function normalize (dataArray: Array<object>, array: number[]): Array<object> {
        const dateToday = new Date();
        const hour = dateToday.getHours();
        for (let i = 12; i >=0; i--) {
            dataArray.push({
                name: ((hour-i)<1 ? 24+hour-i : hour-i) + "h",
                errors: array[i]
            });
        }
        return dataArray;
    }

    const awaitData = async (): Promise<void>=>{
        props.getData().then((dataValueArray)=>{
            let dataArray: Array<object> = [];
            dataArray = normalize(dataArray,dataValueArray);
            setData(dataArray);
        });
    };

    const initalTime = () => {
        // const timeOnMount = new Date().getTime();
        setLastRefreshTime(new Date().getTime());
    }
    useEffect(()=>{
        initalTime();
        console.log(lastRefreshTime + " last refresh time");

        const intervalHandler = (): void => {
            const timeOnInterval = new Date().getTime();
            console.log(timeOnInterval + " time on interval")
            const diffInSeconds = (timeOnInterval-lastRefreshTime)/1000
            console.log(diffInSeconds + " diff");
            const hourToSeconds = 3600;
            if(diffInSeconds>=hourToSeconds){
                setLastRefreshTime(timeOnInterval);
                awaitData();
            }
        }

        const refreshInterval = window.setInterval(intervalHandler, 60000);
        setRefreshIntervalId(refreshInterval);

        awaitData();
    },[])

    return(
        <div className="balance-graph" >
            <div className="graph-header" >
                <div className="graph-title">Error Frequency</div>
                <div className="graph-option position" >IN THE LAST DAY</div>
            </div>
            <BarChart
            width={624} height={199} data={data}
            margin={{top: 5, bottom: 0, left: 10, right: 10,}}>
                <XAxis dataKey="name" stroke="#9ba7af" 
                    interval="preserveStartEnd" tickLine={false}/>
                <Tooltip contentStyle={{color: "red"}} cursor={false} isAnimationActive={false}/>
                <Bar dataKey="errors" fill="#C3CBCF" barSize={28}/>
            </BarChart>
        </div>
    );
}