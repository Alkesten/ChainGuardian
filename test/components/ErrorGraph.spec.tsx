import * as React from "react";
import * as renderer from "react-test-renderer";
import {ErrorGraph} from "../../src/renderer/components/ErrorGraph/ErrorGraph";

const giveData = async (): Promise<number[]> => {
    return [2,5,9,12,1,5,6,8,10,10,11,5,7,8];
};

describe("ErrorGraph", () => {
    it("renders correctly", () => {
        const tree = renderer
            .create(<ErrorGraph 
                getData={giveData}
            />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});