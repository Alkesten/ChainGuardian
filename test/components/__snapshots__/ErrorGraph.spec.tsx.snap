// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorGraph renders correctly 1`] = `
<div
  className="balance-graph"
>
  <div
    className="graph-header"
  >
    <div
      className="graph-title"
    >
      Error Frequency
    </div>
    <div
      className="graph-option position"
    >
      IN THE LAST DAY
    </div>
  </div>
  <div
    className="recharts-wrapper"
    onMouseEnter={[Function]}
    onMouseLeave={[Function]}
    onMouseMove={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    style={
      Object {
        "cursor": "default",
        "height": 199,
        "position": "relative",
        "width": 624,
      }
    }
  >
    <svg
      className="recharts-surface"
      height={199}
      version="1.1"
      viewBox="0 0 624 199"
      width={624}
    >
      <defs>
        <clipPath
          id="recharts2-clip"
        >
          <rect
            height={164}
            width={604}
            x={10}
            y={5}
          />
        </clipPath>
      </defs>
    </svg>
    <div
      className="recharts-tooltip-wrapper"
      style={
        Object {
          "MozTransform": "translate(undefinedpx, undefinedpx)",
          "OTransform": "translate(undefinedpx, undefinedpx)",
          "WebkitTransform": "translate(undefinedpx, undefinedpx)",
          "msTransform": "translate(undefinedpx, undefinedpx)",
          "pointerEvents": "none",
          "position": "absolute",
          "top": 0,
          "transform": "translate(undefinedpx, undefinedpx)",
          "visibility": "hidden",
        }
      }
    >
      <div
        className="recharts-default-tooltip"
        style={
          Object {
            "backgroundColor": "#fff",
            "border": "1px solid #ccc",
            "color": "red",
            "margin": 0,
            "padding": 10,
            "whiteSpace": "nowrap",
          }
        }
      >
        <p
          className="recharts-tooltip-label"
          style={
            Object {
              "margin": 0,
            }
          }
        >
          
        </p>
      </div>
    </div>
  </div>
</div>
`;
